DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,801 @@
+is problematic in situations when the body is missing. In conclusion, `opAssign` can be generated from the
+copy constructor when the `struct` contains only assignable (mutable) fields.
+
+2. The copy constructor signature is: `this(ref $q1 S rhs) $q2`, where `q1` and `q2` represent the
+qualifiers that can be applied to the function and the parameter (`const`, `immutable`, `shared`, etc.). The
+problem that arises is: depending on the values of `$q1` and `$q2` what should the signature of `opAssign` be?
+A solution might be to generate for every copy constructor its counterpart `opAssign`: `void opAssign(ref $q1 S rhs) $q2`.
+However, when is a `const`/`immutable` `opAssign` needed? There might be obscure cases when that is useful, but those are
+niche situations where the user must step it and clarify what the desired outcome is and define its own `opAssign`. For
+the sake of simplicity, `opAssign` will be generated solely for copy constructors that have a missing `$q2`.
+
+Taking into account the above mentioned restrictions we can generate a single `opAssign` function
+and leverage the existing `opAssign` generation logic for the [[postblit](https://github.com/dlang/dmd/pull/8505)].
+Because `opAssing` is generated only if there are copy constructors that create a mutable object, the copy
+constructor is going to be called (if possible) when the source is passed as a parameter to the `opAssign`
+function.
+
+### POD (Plain Old Data)
+
+A struct that defines a copy constructor is not POD.
+
+## Breaking Changes and Deprecations
+
+1. The parameter of the copy constructor is passed by a mutable reference to the
+source object. This means that a call to the copy constructor may legally modify
+the source object:
+
+```d
+struct A
+{
+    int[] a;
+    this(ref A another)
+    {
+        another.a[2] = 3;
+    }
+}
+
+void main()
+{
+    A a, b;
+    a = b;    // b.a[2] is modified
+}
+```
+
+A solution to this might be to make the reference `const`, but that would make code like
+`this.a = another.a` inside the copy constructor illegal. Of course, this can be solved by means
+of casting : `this.a = cast(int[])another.a`.
+
+2. This DIP changes the semantic of constructors which receive a parameter by reference of type
+`typeof(this)`. The consequence is that the constructor might get called implicitly in some
+situations:
+
+```d
+struct C
+{
+    this(ref C another)    // normal constructor before DIP, copy constructor after
+    {
+        import std.stdio : writeln;
+        writeln("Yo");
+    }
+}
+
+void fun(C c) {}
+
+void main()
+{
+    C c;
+    fun(c);
+}
+```
+
+Before this DIP this code would not print anything, while after it will print "Yo". However, a case can be
+made that a constructor with the above definition could not be correctly used as anything else than a copy
+constructor, in which case the current DIP, although surprising, actually fixes the code.
```

Comment:
the current DIP, although surprising, actually fixes the code -> this DIP actually fixes the code.